### PR TITLE
Reduce lock contention in HttpRemoteTask

### PR DESCRIPTION
**Pull request by sopel39**
## Description

Use separate locks for sending a new request
(which is potentially expensive) and modifying
HttpRemoteTask state.

> Is this change a fix, improvement, new feature, refactoring, or other?

improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

core

> How would you describe this change to a non-technical end user or system administrator?

Improve throughput for large clusters with higher concurrency

## Related issues, pull requests, and links

Fixes https://github.com/trinodb/trino/issues/10842

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# General
* Improve throughput for large clusters with higher concurrency. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java [hunk 1]
```diff
@@ -917,12 +929,12 @@ public void success(TaskInfo value)
             try (SetThreadName ignored = new SetThreadName("UpdateResponseHandler-%s", taskId)) {
                 try {
                     long currentRequestStartNanos;
-                    synchronized (HttpRemoteTask.this) {
+                    synchronized (httpClient) {
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java:932]: I do not get this. Now there are places where you modify `currentRequest` without holding `httpClient` lock.

Also with IntelliJ reports lots of:
image


Please enable those inspections:
image

and fix the code so warnings are not reported

File: core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java [hunk 2]
```diff
@@ -126,13 +126,12 @@
...
     // The version of dynamic filters that has been successfully sent to the worker
-    private long sentDynamicFiltersVersion = INITIAL_DYNAMIC_FILTERS_VERSION;
+    private final AtomicLong sentDynamicFiltersVersion = new AtomicLong(INITIAL_DYNAMIC_FILTERS_VERSION);
 
-    @GuardedBy("this")
+    @GuardedBy("httpClient")
     private Future<?> currentRequest;
-    @GuardedBy("this")
+    @GuardedBy("httpClient")
     private long currentRequestStartNanos;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java:135]: I think `currentRequestStartNanos` can be made a member variable of `UpdateResponseHandler` and removed from here, then we don't need locking for it. `TaskStatusResponseCallback` and `DynamicFiltersResponseCallback` already do that

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java:135]: Maybe, but it doesn't really simplify the code here as this is always updated together with `currentRequest`

File: core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java [hunk 3]
```diff
@@ -579,73 +584,75 @@ private void triggerUpdate()
         }
     }
 
-    private synchronized void sendUpdate()
+    private void sendUpdate()
```
Comment (radek-starburst) [inline core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java:587]: If we remove `synchronized` here and from `public void setOutputBuffers(OutputBuffers newOutputBuffers)`
we change the semantic, the methods `sendUpdate` and `setOuputBuffer` are not mutually exclusive. 
In a result we have a situation when the state of "reported" (published by `sendUpdate`) is internally consistent. 

Especially, we could mix values `sendPlan` and `sentDynamicFiltersVersion` in `sendUpdate` from the last one response and from the previous one.

Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java:587]: Yeah - concerns look valid.
@sopel39 what exactly do you want to fix here? Is there a performance bottle neck this change addresses, and is that tested it actually helps. 
If benefit is small I would just leave current simple synchronization. If we really need that the change need to be fine crafted so it is easy to follow (this change is not).

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java:587]: > If we remove synchronized here and from success we change the semantic, the methods sendUpdate and sucess are not mutually exclusive.

`sendUpdate` and `sucess` were never mutually exclusive. However, they still synchronize on `httpClient`, so appropriate fields are guarded.

> In a result we have a situation when the state of "reported" request (published by sendUpdate) is internally inconsistent.

I don't understand that

> Especially, we could mix values sendPlan and sentDynamicFiltersVersion in sendUpdate from the last one response and from the previous one.

It doesn't matter. Both `sentDynamicFiltersVersion` and `sendPlan` are idempotent. `sentDynamicFiltersVersion` never goes down.

> @sopel39 what exactly do you want to fix here? Is there a performance bottle neck this change addresses, and is that tested it actually helps.

Take a look at https://github.com/trinodb/trino/commit/583fcabd51a3757bd4e4ccb4401ebc59e2b5dafc. This issue caused serious regression from release to release. However, that was not a proper fix (rather some "lucky" relaxation of constraints) as all methods are still synchronized on `this` (including expensive `sendUpdate`). This PR splits splits `sendUpdate` synchronization fully

Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java:587]: > sendUpdate and sucess were never mutually exclusive. However, they still synchronize on httpClient, so appropriate fields are guarded.
 
`sentDynamicFiltersVersion` and `sendPlan` are not under `httpClient` lock in the `success` method. 

> I don't understand that
> It doesn't matter. Both sentDynamicFiltersVersion and sendPlan are idempotent. sentDynamicFiltersVersion counter never goes down.

What do you mean they are idempotent? `value.isNeedsPlan` from response (look to the `success` method) cannot be different from `HttpRemoteTask.sendPlan` that is read in `updateTask` from another thread?

**Comment by sopel39** [comment]
I've split it into smaller commits. PTAL @losipiuk

File: core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java (hunk 3 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java:587]: > What do you mean they are idempotent?

It doesn't matter if coordinator sends some DF or plan twice (and `HttpRemoteTask.sendPlan` is true by default). I've changed the code to make sure `sentDynamicFiltersVersion` and `sendPlan` is updated before next request is started (`currentRequest = null`)

File: core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java [hunk 4]
```diff
@@ -581,71 +581,73 @@ private void triggerUpdate()
...
-        // don't update if the task hasn't been started yet or if it is already finished
-        if (!started.get() || taskStatus.getState().isDone()) {
-            return;
-        }
-
-        int currentPendingRequestsCounter = pendingRequestsCounter.get();
-        if (currentPendingRequestsCounter == 0) {
-            return;
-        }
+        synchronized (this) {
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java]: `Add explicit synchronization section` - you can drop the `synchronized` from method alrady.

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java]: no, `sendUpdate` might still be scheduled twice or called from request callback. Only one should execute at same time.

**Review by losipiuk — APPROVED** [review]
Seems fine :/

**Comment by radek-starburst** [comment]
lgtm

**Review by radek-starburst — APPROVED** [review]

**Pull request merged**